### ISSUE TITLE
local file loads in an existing window but variables like "nw", "process", "global" are unavailable

### ISSUE THREAD

**Issue by RohitSharma27** [nwjs/nw.js#4806]
nw0.14.3-SDK
Mac OS X El Capitan
node v5.9.0
I am trying to port my application from nw0.12.3 to nw0.14.3.
I am observing a very strange behaviour. I have got this file that is injected via "inject_js_start". In that file i have "require()" statements. When I fire my app, I am getting an error "require is not defined". I try to open devtool for backgound page and wait for a while, the error goes away sometimes and sometimes it don't. But, if I replace "require()" with "nw.require()" then the error goes away.
Why is this happening?
After this I am getting the subjected error. Below are the screenshots. What could be wrong in here?
<img width="641" alt="screen shot 2016-05-03 at 5 49 12 pm" src="https://cloud.githubusercontent.com/assets/18210507/14982915/911bab70-1157-11e6-8b7d-81a254fddbad.png">
<img width="641" alt="screen shot 2016-05-03 at 5 50 00 pm" src="https://cloud.githubusercontent.com/assets/18210507/14982925/a20f2830-1157-11e6-88f5-9c5823621ae2.png">

**Comment by frankhale** [comment]
Yeah I'm seeing this as well. I put some rather simple code in a module that I require in my index.html page. That code gets the current window, sets the window on loaded event and opens the dev tools. I can't do it from that module because of this error. If I put that code directly in my index.html it works fine. This seems like a bug to me. I can make a reproducible test case if need be.

**Comment by frankhale** [comment]
Yeah I'm seeing this in a simple app. I wanted to take the code in my index page and put it in a simple node module then require it on my index page. In that code I was attempting to get the current window, set the window on loaded event and then open the dev tools. This won't work because of the above error. If I take the same code and copy it into a script tag in the index page it works fine. Is it not possible to require nw.gui in the simple node module and do what I want? It doesn't seem possible right now.

I'm using 0.15 beta2 x64 on Windows 10

**Comment by frankhale** [comment]
Sorry to pollute this discussion here. I think the following best describes my issue, I'm getting the JS contexts mixed up.

> You cannot use require('nw.gui') (to access the node-webkit's GUI API) from the Node's context, because there's no GUI outside of a window.

**Comment by RohitSharma27** [comment]
@frankhale No, you can't do that in node context. And this isn't the case with me. I am including the file via "inject_js_start" and performing "require()" in that. I get the message 
`Uncaught ReferenceError: require is not defined`
because I am assigning it to an object like

> nodeApp.lib = require('nw.gui');

If I replace it by "nw.require('nw.gui')" then this error goes away. Now my code executes a little further and I see the above subjected error that I have attached as an image. I am trying to know what could have caused this behaviour? I am really on a tight deadline here and I don't know where to start looking.

**Comment by RohitSharma27** [comment]
I have narrowed down my findings on the app that i am trying to port on nw0.14.3 and here is what is happening.
1. "require()" works fine all the time.
2. I try to load a local html file in the main app window after sometime using 
`nw.Window.get(window.location.assign('/Users/rohit/Library/Application\ Support/TestApp/Default/data/native/res/html/page.html?app=HTML5Desktop'))` 
and I get "Your file was not found" error.
File is there, I assure you. When it tries to load this html page, window reloads and "app.js" is executed again which is the desired behaviour but this time "require is undefined". 
Infact, there is no "process" object, no "nw" object available in console. 
If I write 

> location.href
I get `data:text/html,chromewebdata`, because the page was never loaded.

**Comment by RohitSharma27** [comment]
So, now I can load a local file in the current window. All I had to do was prepend "file:///"  in the url I want to load. So, this is how my code looks like 
`nw.Window.get(window.location.assign('file:///Users/rohit/Library/Application\ Support/TestApp/Default/data/native/res/html/page.html?app=HTML5Desktop'))`
The page loads fine but now there is no "nw" and "process" object available in the window.
This is a blocker for me.
In nw0.12.3 it was working fine as nw was using "file://" protocol to load local files. But, now if I don't prepend "file:///" in my url then nw tries to access using 
`chrome-extension://<chrome.runtime.id>/<path to the file>` which will never be there.

**Comment by RohitSharma27** [comment]
@rogerwang Please suggest any alternative to this. This is a real blocker for me.
PS : My package.json's node-remote has "<all_ulrs>" value.

**Comment by RohitSharma27** [comment]
@ghostoy Can you at least guide me in the right direction, Please? It would be highly appreciated.
Thanks.

**Comment by siddhartha-gupta** [comment]
Please attach a demo app for debugging purpose

**Comment by RohitSharma27** [comment]
Please download the package.nw directory from [https://www.dropbox.com/s/wq5rok4g0q7t13i/package.nw.zip?dl=0](url) and place the unzipped directory along with your "nw.exe". There is a "local.html" file in the "package.nw" directory, place it in your "Roaming" directory if you're using windows machine, or you may place it anywhere and edit the "app.js" file accordingly.
I have tried it with nwv0.14.5 SDK build on windows 7. The issue is still there.
Thanks.

**Comment by siddhartha-gupta** [comment]
Try adding following chromium arguments in package.json

"chromium-args": "--allow-file-access-from-files --allow-file-access --user-data-dir --mixed-context"

**Comment by RohitSharma27** [comment]
No, it isn't helping. The problem is still there.

**Comment by Rishu** [comment]
I am facing similar problem due to that i am unable to move ahead. Please provide us any workaround to get it fix.

Thanks in advance.

**Comment by narendrag1** [comment]
For accessing local file use : chrome-extension protocol rather than file protocol.
For me, it  was not clear from 0.12 to 0.13 migration notes how to get chrome extension app id : 

You can get chrome-id as :  chrome.runtime.id

Here is how I created new nw window process pointing to local file. And was able to access : nw, require, process etc.

I added these below lines of code to my app main file ( which is the html file or js file that is mentioned as "main": "appMain.html"  in package.json file in app folder of your application.

var newWindowOptions = {
			"new_instance": true,
			"show": true
		};
    
     	var newWin= nw.Window.open('chrome-extension://' + chrome.runtime.id + '/path_relative_to_your_main_file_to_ newwin.html', newWindowOptions);

**Comment by RohitSharma27** [comment]
@narendrag1 Thanks for suggesting this approach. This is a work around.
But, using this approach to access local files is not good for an application. This would be like hard coding the urls, as I'll have to go back relative to my main html file(which depends on the directory structure used by the chromium) and then I'll be traversing the directory structures accordingly. However, the "file:///" protocol still needs to be implemented properly in  nwjs.
By the way, your `newWin` will always be `undefined`. `Window.open()` provides the Window object in the callback function. It doesn't return it anymore. :)

**Comment by asinverse** [comment]
I have the same problem too. 
I tried @narendrag1's approach, but the relative path is `/../../../local/index.html` and NW removes all `/../../..`, then I get final path `chrome-extension://opibfileoidelifflmbocplppfckeegn/local/index.html` and error `Your file was not found`.

For info, I'm working on NW v.0.14.6 and also set: `"chromium-args": "--allow-file-access-from-files --allow-file-access"` and `"node-remote": ["<all_urls>"]` in the package.json

**Comment by RohitSharma27** [comment]
Any update on this @rogerwang ? I need to access a local file and need both node and nw objects. I have got two options at this point of time. As much as I understand, `chrome-extension://` protocol can't access anything outside the `package.nw` directory. If I try to access any local file then `file://` protocol is used and there is no node or nw object present in this case.There is this last option of running a node server and serve files through that, but this would bring some complications in my code base.
Please suggest any alternative.

**Comment by linb** [comment]
Now, any solution available about this?

**Comment by RohitSharma27** [comment]
No, if you want to access some file using file protocol.

**Comment by Christywl** [comment]
Could you please share a sample to reproduce this issue?

**Comment by RohitSharma27** [comment]
Hi christywl, 
I will upload a reproducible case as soon as I can. You can easily reproduce this loading a URL in a window using the file protocol I.e. let us suppose you have some HTML file placed in some directory on your machine. So, just load it in a window using file:///path to HTMLfile. Now open the dev tool and try to access the above mentioned objects. They won't be accessible.

**Comment by Christywl** [comment]
I use a simple code to test, `nw`/`process`/`global` work fine on Windows/Mac `dev tools/console`with nwjs-sdk-v0.21.6.
```
<!DOCTYPE html>
<html>
<body>
<script>
var win = nw.Window.open('file:///C:/nwjs/local.html',{}, function(new_win) {
  console.log('new window');
});
</script>
</body>
</html>
```
Similar with #5320 ?

If my code is wrong, I'll wait for your sample to check again.

**Comment by Christywl** [comment]
close until there is more information

**Comment by jkfb** [comment]
This issue still exits. I can reproduce it, see information below.
NW version: 0.36.1-sdk
Platform: Mac, OS X Mojave, Version 10.14.4

package.json:
```
{
    "name": "nw-demo",
    "main": "index.html",
    "node-remote": "<all_urls>"
}
```
index.html
```
<body>
  <button onclick="goto();">
    click me!
  </button>
  <button onclick="goto2();">
    file me!
  </button>
  <button onclick="goto3();">
    open me!
  </button>
  <script>
    var goto = function goto() {
      // works fine!
      location.href="./main.html";
    };

    var goto2 = function goto2() {
      // 'nw' will be 'undefined' in 'main.html' 
      location.href ="file:///" + nw.global.__dirname + "/main.html";
    };

    var goto3 = function goto3() {
      // 'nw' will be 'undefined' in 'main.html' 
      var win = nw.Window.open("file:///" + nw.global.__dirname + "/main.html",{} , function(new_win) {
        console.log('new window');
      });
    };
  </script>
</body>
```
main.html
```
<body>
  <script>
    document.write(nw.process.version);
  </script>
</body>
```

**Comment by Hudell** [comment]
Any news about this? Still running into this problem on nwjs-sdk-v0.33.3.
Even with node-remote set to `<all_urls>` and the simple code shared by @Christywl, `nw`, `process` and `require` are not available on the new window.
